### PR TITLE
Add surface syntax for dependent pairs

### LINKED ISSUE #966: Spell dependent pairs in the surface syntax
There are perfectly good `DepPair` and `DepPairTy` constructors for `Atom`s, but `UExpr` doesn't seem to be able to represent any syntax that would parse thereto?  Would be convenient at least for `transpose_lower_ix` and friends being introduced in #963.

### PR DESCRIPTION
**Pull request by axch**
Fixes #966.

Co-authored-by: Dougal Maclaurin <dougalm@google.com>

### CODE REVIEW

File: src/lib/Inference.hs [hunk 1]
```diff
@@ -964,15 +964,19 @@ checkOrInferRho (WithSrcE pos expr) reqTy = do
...
+      -- it would get bound to the inserted arguments instead of the
+      -- desired dict. It's not a fundemental limitation of our
+      -- automatic quantification, but it's simpler not to deal with
       -- that for now.
-      let checkNoMissing = addSrcContext pos' $ unless (null $ eSetToList missingDs) $ throw TypeErr $
-            "Couldn't synthesize a class dictionary for: " ++ pprint (head $ eSetToList missingDs)
+      let checkNoMissing = (addSrcContext pos'
+           $ unless (null $ eSetToList missingDs) $ throw TypeErr
+           $ "Couldn't synthesize a class dictionary for: "
+             ++ pprint (head $ eSetToList missingDs))
```
Comment (apaszke) [inline src/lib/Inference.hs:979]: Please avoid reformatting if you're not actually modifying the code. Or split the reformatting changes into a separate commit

File: tests/type-tests.dx [hunk 2]
```diff
@@ -350,13 +350,16 @@ def triRefIndex {h n} (ref:Ref h ((i':n)=>(..i')=>Float)) (i:n) : Ref h ((..i)=>
...
+-- the checking doesn't provide any insight into what the argument is
+-- in this case. This checks that type inference is able to realize
+-- that and switch to inference mode, so that it can correctly infer
+-- the full dependent type.
 --
--- There was a time when this wasn't possible, because checking mode would unify the
--- input type with a non-dependent function type, leading to a later unification errors.
+-- There was a time when this wasn't possible, because checking mode
+-- would unify the input type with a non-dependent function type,
+-- leading to a later unification errors.
```
Comment (apaszke) [inline tests/type-tests.dx:361]: Again, please don't format things unnecessarily

File: src/lib/Inference.hs [hunk 3]
```diff
@@ -1008,6 +1012,26 @@ checkOrInferRho (WithSrcE pos expr) reqTy = do
...
+    ann' <- checkAnn ann
+    depPairTy <- addSrcContext pos' do
+      buildDepPairTyInf (getNameHint pat) ann' \v -> do
+        Abs decls result <- buildDeclsInf do
+          v' <- sinkM v
+          bindLamPat (WithSrcB pos' pat) v' (checkUType rhs)
+        cheapReduceWithDecls decls result >>= \case
+          (Just rhs', DictTypeHoistSuccess, []) -> return rhs'
+          _ -> throw TypeErr $ "Can't reduce type expression: " ++
+                 docAsStr (prettyBlock decls result)
```
Comment (apaszke) [inline src/lib/Inference.hs:1025]: Is this any different from `checkUType`?

**Review by apaszke — COMMENTED** [review]
Can you add some more tests that actually evaluate some expressions using dependent pairs? Perhaps have a case statement that returns lists of different lengths? And make sure we can print those.

File: src/lib/Inference.hs (hunk 1 above)
Comment (axch) [reply src/lib/Inference.hs:979]: Split into separate commit.

**Comment by axch** [comment]
Also added a test the involves going through a case, and printing out a dependent pair.  Added a clause for dependent pairs to `Serialize.hs` to make them print more nicely than the default.

File: tests/type-tests.dx [hunk 4]
```diff
@@ -520,6 +520,24 @@ foo : Fin 3 => Float = [1.0, 2.0, 3.0]
...
+
+:t pair
+> (n:Nat &> ((Fin n) => Nat))
+
+'Check that dependent pairs typecheck across a case, and that they print sensibly.
+
+if False
+  then (5 ,> buffer) :: tp
+  else (3 ,> for i:(Fin 3). (ordinal i) * 2) :: tp
+> (3 ,> [0, 2, 4]) @ (n:Nat &> ((Fin n) => Nat))
```
Comment (apaszke) [inline tests/type-tests.dx:539]: The `@` printing is a little unfortunate, since it is an operator we use for index casting. Perhaps use the `::` syntax?

In the long term, we should consider replacing `::` with `@`, since that might be less confusing. We would need to find another operator for index casts, though.

**Review by apaszke — APPROVED** [review]

File: tests/type-tests.dx (hunk 4 above)
Comment (axch) [reply tests/type-tests.dx]: Done.
